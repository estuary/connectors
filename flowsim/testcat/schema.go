package testcat

import (
	"reflect"
	"strings"
	"time"
)

// CustomSchema is an interface that allows a struct to build it's own schema
// rather than use the default schema builder.
type CustomBuildSchema interface {
	BuildSchema() (TestSchema, []string, error)
}

// BuildSchema parses a struct and builds the flow.yaml schema representation along with it's keys.
// It can handle just about any struct. To determine the name of the fields in the schema, you can use the
// flowsim tag to set the name. You can also use the 'key' tag option to set the struct field as a key and
// 'required' to set it as a required field. (keys are automatically required) See testdata.Basic for
// and example struct.
func BuildSchema(in interface{}) (TestSchema, []string, error) {

	// If it supports the CustomBuildSchema interface, use it.
	if cbs, ok := in.(CustomBuildSchema); ok {
		return cbs.BuildSchema()
	}

	var schema TestSchema
	var keys []string
	var err error

	// Handle any special built-in types such as time.Time.
	switch in.(type) {
	case time.Time:
		schema.Type = []string{"string"}
		schema.Format = "date-time"
		return schema, keys, err
	}

	value := reflect.ValueOf(in)
	valueKind := value.Kind()

	// Dereference if a pointer.
	if (valueKind == reflect.Ptr && !value.IsNil()) || valueKind == reflect.Interface {
		value = value.Elem()
		valueKind = value.Kind()
	}

	switch valueKind {
	// Support maps, but it won't have any keys.
	case reflect.Map:
		schema.Type = []string{"object"}
		schema.Properties = make(map[string]TestSchema)
		for _, mapKey := range value.MapKeys() {
			if schema.Properties[mapKey.String()], _, err = BuildSchema(value.MapIndex(mapKey).Interface()); err != nil {
				return schema, keys, err
			}
		}
	// For structs, parse all the fields.
	case reflect.Struct:
		schema.Type = []string{"object"}
		schema.Properties = make(map[string]TestSchema)
		for i := 0; i < value.NumField(); i++ {
			structKey := value.Type().Field(i)

			// Non exported fields should be skipped.
			if structKey.PkgPath != "" {
				continue
			}

			// Parse the flowsim tags to find required fields and keys.
			fieldName := structKey.Tag.Get("flowsim")
			tagOptions := []string{}
			if fieldName == "" {
				fieldName = structKey.Name
			} else {
				// Get the field name and tagOptions.
				if idx := strings.Index(fieldName, ","); idx != -1 {
					tagOptions = strings.Split(fieldName[idx+1:], ",")
					fieldName = fieldName[:idx]
				}
				// Skipped field.
				if fieldName == "-" {
					continue
				}
			}
			// If it's a key, mark it as key and as required.
			if optionsHas(tagOptions, "key") {
				keys = append(keys, "/"+fieldName)
				schema.Required = append(schema.Required, fieldName)
			} else if optionsHas(tagOptions, "required") {
				// Mark as required
				schema.Required = append(schema.Required, fieldName)
			}
			// Parse the field properties (as well as child objects).
			var fieldKeys []string
			if schema.Properties[fieldName], fieldKeys, err = BuildSchema(value.Field(i).Interface()); err != nil {
				return schema, keys, err
			}
			// Any child keys need to be appended to the keys slice.
			for _, key := range fieldKeys {
				keys = append(keys, "/"+fieldName+key)
			}
		}
	default:
		// All other types, use reflect.Kind.
		schema.Type = SchemaType(valueKind)
	}

	return schema, keys, nil

}

// optionHas is a helper to parse struct tags and see if it has an option.
func optionsHas(options []string, value string) bool {
	for _, s := range options {
		if s == value {
			return true
		}
	}
	return false
}

// SchemaType returns the json type for a field.
func SchemaType(t reflect.Kind) []string {

	switch t {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return []string{"integer"}
	case reflect.Array, reflect.Slice:
		return []string{"array"}
	case reflect.Float32, reflect.Float64:
		return []string{"number"}
	case reflect.Bool:
		return []string{"boolean"}
	case reflect.String:
		return []string{"string"}
	}
	return []string{"unknown"}

}

// BuildCollection is a helper to automatically build a schema, parse keys and return a Collection.
func BuildCollection(in interface{}) (TestCollection, error) {
	var c TestCollection
	var err error
	if c.Schema, c.Keys, err = BuildSchema(in); err != nil {
		return c, err
	}
	return c, err
}
