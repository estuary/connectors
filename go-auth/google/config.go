package google_auth

import (
	"context"
	"fmt"
	"time"

	"github.com/iancoleman/orderedmap"
	"github.com/invopop/jsonschema"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

const (
	SERVICE_AUTH_TYPE = "Service" // For service account credentials
	CLIENT_AUTH_TYPE  = "Client"  // For clients (web browsers) to use oauth
)

type CredentialConfig struct {
	AuthType        string `json:"auth_type"`
	ClientID        string `json:"client_id,omitempty"`
	ClientSecret    string `json:"client_secret,omitempty"`
	AccessToken     string `json:"access_token,omitempty"`
	RefreshToken    string `json:"refresh_token,omitempty"`
	CredentialsJSON string `json:"credentials_json,omitempty"`
}

func (c *CredentialConfig) Validate() error {
	switch c.AuthType {
	case SERVICE_AUTH_TYPE:
		return c.validateServiceCreds()
	case CLIENT_AUTH_TYPE:
		return c.validateClientCreds()
	default:
		return fmt.Errorf("invalid credentials auth type %q", c.AuthType)
	}
}

func (c *CredentialConfig) validateServiceCreds() error {
	if c.CredentialsJSON == "" {
		return fmt.Errorf("missing service account credentials JSON")
	}

	return nil
}

func (c *CredentialConfig) validateClientCreds() error {
	if c.RefreshToken == "" {
		return fmt.Errorf("missing refresh token for oauth2")
	} else if c.ClientID == "" {
		return fmt.Errorf("missing client ID for oauth2")
	} else if c.ClientSecret == "" {
		return fmt.Errorf("missing client secret for oauth2")
	}

	return nil
}

func (c *CredentialConfig) GoogleCredentials(ctx context.Context, scopes ...string) (*google.Credentials, error) {
	if err := c.Validate(); err != nil {
		return nil, err
	}

	if c.AuthType == SERVICE_AUTH_TYPE {
		return google.CredentialsFromJSON(ctx, []byte(c.CredentialsJSON), scopes...)
	}

	oauthConfig := &oauth2.Config{
		ClientID:     c.ClientID,
		ClientSecret: c.ClientSecret,
		Endpoint:     google.Endpoint,
		Scopes:       scopes,
	}

	token := &oauth2.Token{
		RefreshToken: c.RefreshToken,
		Expiry:       time.Now(), // Require a new access token to be obtained.
	}
	tokenSource := oauthConfig.TokenSource(ctx, token)

	return &google.Credentials{
		TokenSource: tokenSource,
	}, nil
}

// JSONSchema allows for the schema to be (semi-)manually specified when used with the
// github.com/invopop/jsonschema package in go-schema-gen, to fullfill the required schema shape for
// our oauth connectors.
func (CredentialConfig) JSONSchema() *jsonschema.Schema {
	serviceAccountProps := orderedmap.New()
	serviceAccountProps.Set("auth_type", &jsonschema.Schema{
		Type:    "string",
		Default: SERVICE_AUTH_TYPE,
		Const:   SERVICE_AUTH_TYPE,
	})
	serviceAccountProps.Set("credentials_json", &jsonschema.Schema{
		Title:       "Service Account JSON",
		Description: "The JSON key of the service account to use for authorization.",
		Type:        "string",
		Examples: []interface{}{
			"{ \"type\": \"service_account\" \"project_id\": YOUR_PROJECT_ID, \"private_key_id\": YOUR_PRIVATE_KEY, ... }",
		},
		Extras: map[string]interface{}{
			"secret":    true,
			"multiline": true,
		},
	})

	oauthProps := orderedmap.New()
	oauthProps.Set("auth_type", &jsonschema.Schema{
		Type:    "string",
		Default: CLIENT_AUTH_TYPE,
		Const:   CLIENT_AUTH_TYPE,
	})
	oauthProps.Set("client_id", &jsonschema.Schema{
		Type: "string",
		Extras: map[string]interface{}{
			"secret": true,
		},
	})
	oauthProps.Set("client_secret", &jsonschema.Schema{
		Type: "string",
		Extras: map[string]interface{}{
			"secret": true,
		},
	})
	oauthProps.Set("access_token", &jsonschema.Schema{
		Type: "string",
		Extras: map[string]interface{}{
			"secret": true,
		},
	})
	oauthProps.Set("refresh_token", &jsonschema.Schema{
		Type: "string",
		Extras: map[string]interface{}{
			"secret": true,
		},
	})

	return &jsonschema.Schema{
		Title:       "Authentication",
		Description: "Google API Credentials",
		Default:     map[string]string{"auth_type": CLIENT_AUTH_TYPE},
		OneOf: []*jsonschema.Schema{
			{
				Title:      "Google OAuth",
				Required:   []string{"auth_type", "client_id", "client_secret", "refresh_token"},
				Properties: oauthProps,
				Extras: map[string]interface{}{
					"x-oauth2-provider": "google",
				},
			},
			{
				Title:      "Service Account Key",
				Required:   []string{"auth_type", "credentials_json"},
				Properties: serviceAccountProps,
			},
		},
		Extras: map[string]interface{}{
			"discriminator": map[string]string{"propertyName": "auth_type"},
		},
		Type: "object",
	}
}
