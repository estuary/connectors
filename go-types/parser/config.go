package parser

import (
	"encoding/json"
	"os"
)

type JsonPointer string

// Config represents the parser configuration json. It matches the Rust type used by the parser.
// Eventually, it might be nice to generate this definition from the json schema output from the
// parser.
type Config struct {
	AddSourceOffset       string                      `json:"addSourceOffset,omitempty"`
	AddValues             map[JsonPointer]interface{} `json:"addValues,omitempty"`
	Format                string                      `json:"format,omitempty"`
	Filename              string                      `json:"filename,omitempty"`
	ContentType           string                      `json:"contentType,omitempty"`
	ContentEncoding       string                      `json:"contentEncoding,omitempty"`
	Projections           map[string]JsonPointer      `json:"projections,omitempty"`
	Schema                interface{}                 `json:"schema,omitempty"`
	FileExtensionMappings map[string]string           `json:"fileExtensionMappings,omitempty"`
}

func (c *Config) Copy() Config {
	var newAddValues = make(map[JsonPointer]interface{})
	for k, v := range c.AddValues {
		newAddValues[k] = v
	}
	var newProjections = make(map[string]JsonPointer)
	for k, v := range c.Projections {
		newProjections[k] = v
	}
	var newFileMappings = make(map[string]string)
	for k, v := range c.FileExtensionMappings {
		newFileMappings[k] = v
	}

	return Config{
		AddSourceOffset: c.AddSourceOffset,
		AddValues:       newAddValues,
		Format:          c.Format,
		Filename:        c.Filename,
		ContentType:     c.ContentType,
		ContentEncoding: c.ContentEncoding,
		Projections:     newProjections,
		// TODO: Figure out a non-heinous way to copy an interface{}
		Schema:                c.Schema,
		FileExtensionMappings: newFileMappings,
	}
}

// WriteTo writes the Config as JSON to a file at the given path. The file will be created if it
// does not exist, and will be overwritten if it does exist.
func (c *Config) WriteTo(path string) error {
	var file, err = os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0640)
	if err != nil {
		return err
	}
	return c.WriteToFile(file)
}

// WriteToFile writes the config to the given open file, which must be writable.
func (c *Config) WriteToFile(file *os.File) error {
	var err = json.NewEncoder(file).Encode(c)
	var ioerr = file.Close()
	if err == nil {
		return ioerr
	}
	return err
}
