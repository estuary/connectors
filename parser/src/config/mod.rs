pub mod csv;

use encoding_rs::Encoding;
use schemars::{gen, schema as schemagen};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;
use std::convert::TryFrom;
use std::fmt;
use std::fs;
use std::io;
use std::path::Path;

use self::csv::CharacterSeparatedConfig;

/// References an encoding by WHATWG label.
#[derive(Debug, PartialEq, Copy, Clone, Serialize, Deserialize)]
pub struct EncodingRef(&'static Encoding);

impl schemars::JsonSchema for EncodingRef {
    fn schema_name() -> String {
        "encoding".to_string()
    }
    fn json_schema(_gen: &mut gen::SchemaGenerator) -> schemagen::Schema {
        serde_json::from_value(serde_json::json!({
            "title": "encoding",
            "description": "An encoding scheme, identified by WHATWG label. The list of allowable values is available at: https://encoding.spec.whatwg.org/#names-and-labels",
            "type": "string",
            "pattern": "^[a-z0-9_\\-:]{2,20}$"
        })).unwrap()
    }
}

impl EncodingRef {
    /// Returns the actual encoding_rs struct reference. This function signature may change
    /// if we want to add support for other encodings beyond what's provided by encoding_rs.
    pub(crate) fn encoding(&self) -> &'static Encoding {
        self.0
    }

    pub fn is_utf8(&self) -> bool {
        self.encoding() == encoding_rs::UTF_8
    }
}

impl From<&'static Encoding> for EncodingRef {
    fn from(e: &'static Encoding) -> EncodingRef {
        EncodingRef(e)
    }
}

impl<'a> TryFrom<&'a str> for EncodingRef {
    type Error = String;

    fn try_from(s: &'a str) -> Result<Self, Self::Error> {
        if let Some(encoding) = Encoding::for_label_no_replacement(s.as_bytes()) {
            Ok(EncodingRef(encoding))
        } else {
            Err(format!("no such WHATWG encoding label: '{}'", s))
        }
    }
}

#[derive(
    Eq, PartialEq, Hash, PartialOrd, Ord, Clone, Debug, schemars::JsonSchema, Serialize, Deserialize,
)]
#[schemars(example = "JsonPointer::example")]
pub struct JsonPointer(#[schemars(schema_with = "JsonPointer::schema")] pub String);

impl AsRef<str> for JsonPointer {
    fn as_ref(&self) -> &str {
        self.0.as_str()
    }
}

impl<T: Into<String>> From<T> for JsonPointer {
    fn from(s: T) -> Self {
        JsonPointer(s.into())
    }
}

impl JsonPointer {
    pub fn example() -> Self {
        JsonPointer("/json/pointer".to_string())
    }
    fn schema(_: &mut gen::SchemaGenerator) -> schemagen::Schema {
        serde_json::from_value(serde_json::json!({
            "type": "string",
            "pattern": "^(/[^/]+)*$",
        }))
        .unwrap()
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(try_from = "String", into = "String")]
pub enum Format {
    /// JSON objects separated by whitespace, typically a single newline. This format works for
    /// JSONL (a.k.a. JSON-newline), but also for any stream of JSON objects, as long as they have
    /// at least one character of whitespace between them.
    Json,
    /// Comma-separated values
    Csv,
    /// Tab-separated values
    Tsv,
}

impl std::convert::TryFrom<String> for Format {
    type Error = String;
    fn try_from(s: String) -> Result<Self, Self::Error> {
        s.parse()
    }
}

impl Into<String> for Format {
    fn into(self) -> String {
        self.id().to_string()
    }
}

impl schemars::JsonSchema for Format {
    fn schema_name() -> String {
        "format".to_string()
    }
    fn json_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        let possible_values = Format::all().iter().map(Format::id).collect::<Vec<_>>();
        serde_json::from_value(serde_json::json!({
            "type": "string",
            "enum": possible_values,
            "title": "format",
            "description": "Specifies the format to use for parsing, which overrides the default behavior of infering the format.",
        }))
        .unwrap()
    }
}

impl Format {
    pub fn id(&self) -> &'static str {
        match *self {
            Format::Json => "json",
            Format::Csv => "csv",
            Format::Tsv => "tsv",
        }
    }

    pub fn all() -> &'static [Format] {
        &[Format::Json, Format::Csv, Format::Tsv]
    }
}
impl fmt::Display for Format {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.id())
    }
}

impl std::str::FromStr for Format {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        for format in Format::all() {
            if format.id() == s {
                return Ok(*format);
            }
        }
        Err(format!("invalid format id: '{}'", s))
    }
}

#[derive(Clone, PartialEq, Debug, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ParseConfig {
    /// format forces the use of the given parser and disables automatic format detection. If
    /// unspecified, then the format will be inferred from the filename, content-type, or file
    /// contents.
    #[serde(default)]
    pub format: Option<Format>,

    /// filename is used for format inference. It will be ignored if `format` is specified.
    #[serde(default)]
    pub filename: Option<String>,

    /// The mime type of the data, if known. This will be used for format inference, or ignored if
    /// `format` is specified.
    #[serde(default)]
    pub content_type: Option<String>,

    #[serde(default)]
    /// Add the record offset as a property of each object at the location given. The offset is a
    /// monotonic counter that starts at 0 and increases by 1 for each output document.
    pub add_record_offset: Option<JsonPointer>,

    /// Static data to add to each output JSON document.
    #[serde(default)]
    pub add_values: BTreeMap<JsonPointer, Value>,

    /// Projections control how tabular data like CSV gets transformed into potentially nested JSON
    /// structures. The keys are field names, which may match column names in the source data, and
    /// the values are json pointers indicating where to place the values within the output JSON
    /// document.
    #[serde(default)]
    pub projections: BTreeMap<String, JsonPointer>,

    /// JSON schema describing the desired shape of the output JSON documents. Output documents
    /// will not be validated against this schema, but it can be used to automatically infer
    /// projections for mapping tabular data to nested JSON structures.
    #[serde(default)]
    pub schema: Value,

    /// Mappings from file extensions to format identifiers.
    #[serde(default)]
    pub file_extension_mappings: BTreeMap<String, Format>,

    /// Mappings from content types to format identifiers.
    #[serde(default)]
    pub content_type_mappings: BTreeMap<String, Format>,

    // Below are format-specific configurations, which are used by the parsers.
    /// Configures handling of comma-separated values (CSV) format.
    #[serde(default)]
    pub csv: Option<CharacterSeparatedConfig>,

    /// Configures handling of tab-separated values (TSV) format.
    #[serde(default)]
    pub tsv: Option<CharacterSeparatedConfig>,
}

#[derive(Debug, thiserror::Error)]
pub enum ConfigError {
    #[error("failed to read config: {0}")]
    Io(#[from] io::Error),

    #[error("failed to parse config: {0}")]
    Parse(#[from] serde_json::Error),
}

impl ParseConfig {
    /// Returns the generated json schema for the configuration file.
    pub fn json_schema() -> schemars::schema::RootSchema {
        let mut settings = schemars::gen::SchemaSettings::draft07();
        settings.option_add_null_type = false;
        let generator = schemars::gen::SchemaGenerator::new(settings);
        generator.into_root_schema_for::<ParseConfig>()
    }

    pub fn load(path: impl AsRef<Path>) -> Result<ParseConfig, ConfigError> {
        let file = fs::File::open(path)?;
        let conf = serde_json::from_reader(io::BufReader::new(file))?;
        Ok(conf)
    }

    pub fn override_from(mut self, other: &ParseConfig) -> Self {
        if other.format.is_some() {
            self.format = other.format.clone();
        }
        if other.filename.is_some() {
            self.filename = other.filename.clone();
        }
        if other.content_type.is_some() {
            self.content_type = other.content_type.clone();
        }
        if other.add_record_offset.is_some() {
            self.add_record_offset = other.add_record_offset.clone();
        }
        self.add_values.extend(
            other
                .add_values
                .iter()
                .map(|kv| (kv.0.clone(), kv.1.clone())),
        );
        self.projections.extend(
            other
                .projections
                .iter()
                .map(|kv| (kv.0.clone(), kv.1.clone())),
        );
        if other.schema != Value::Null {
            self.schema = other.schema.clone();
        }
        self.file_extension_mappings.extend(
            other
                .file_extension_mappings
                .iter()
                .map(|kv| (kv.0.clone(), kv.1.clone())),
        );
        self
    }

    pub fn with_format(mut self, format: Format) -> Self {
        self.format = Some(format);
        self
    }
}

impl Default for ParseConfig {
    fn default() -> Self {
        ParseConfig {
            format: None,
            filename: None,
            content_type: None,
            add_record_offset: None,
            add_values: BTreeMap::new(),
            projections: BTreeMap::new(),
            schema: Value::Null,
            file_extension_mappings: default_file_extension_mappings(),
            content_type_mappings: default_content_type_mappings(),
            csv: None,
            tsv: None,
        }
    }
}

fn default_content_type_mappings() -> BTreeMap<String, Format> {
    (&[
        ("application/json", Format::Json),
        ("text/json", Format::Json),
        ("text/csv", Format::Csv),
        ("text/tab-separated-values", Format::Tsv),
    ])
        .iter()
        .map(|(k, v)| (k.to_string(), *v))
        .collect()
}

fn default_file_extension_mappings() -> BTreeMap<String, Format> {
    (&[
        ("jsonl", Format::Json),
        ("json", Format::Json),
        ("csv", Format::Csv),
        ("tsv", Format::Tsv),
    ])
        .iter()
        .map(|(k, v)| (k.to_string(), *v))
        .collect()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn config_schema_is_generated() {
        let schema = ParseConfig::json_schema();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn test_override_from() {
        let base: ParseConfig = serde_json::from_str(
            r#"{
            "format": "json",
            "contentType": "nooooo",
            "addValues": {
                "/foo": "bar",
                "/baz": 2
            },
            "projections": {
                "weee": "wooo"
            }
        }"#,
        )
        .unwrap();
        let actual = base.override_from(
            &serde_json::from_str(
                r#"{
                    "contentType": "application/json",
                    "addValues": {
                        "/foo": "newFoo",
                        "/new": "new"
                    },
                    "schema": true,
                    "addRecordOffset": "/offset",
                    "projections": {
                        "fee": "fi"
                    }
                }"#,
            )
            .unwrap(),
        );

        let expected: ParseConfig = serde_json::from_str(
            r#"{
                "format": "json",
                "contentType": "application/json",
                "addValues": {
                    "/baz": 2,
                    "/foo": "newFoo",
                    "/new": "new"
                },
                "schema": true,
                "addRecordOffset": "/offset",
                "projections": {
                    "weee": "wooo",
                    "fee": "fi"
                }
            }"#,
        )
        .unwrap();
        assert_eq!(expected, actual);
    }
}
