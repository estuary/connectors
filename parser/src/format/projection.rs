//! Types for reasoning about projections of tabular data into potentially nested JSON.
use crate::config::ParseConfig;
use doc::inference::{Exists, Shape};
use doc::{Pointer, Schema, SchemaIndex};
use json::schema::build::Error as SchemaBuildError;
use json::schema::index::Error as SchemaIndexError;
use json::schema::types;
use serde_json::Value;
use std::collections::BTreeMap;

/// Information known about a specific location within a JSON document.
#[derive(Debug, Clone, PartialEq)]
pub struct TypeInfo {
    /// The possible JSON types for this location, if any type information could be inferred.
    pub possible_types: Option<types::Set>,
    pub must_exist: bool,
    pub target_location: Pointer,
}

impl TypeInfo {
    fn from_shape(target_location: Pointer, shape: &Shape, exists: Exists) -> TypeInfo {
        TypeInfo {
            target_location,
            must_exist: exists == Exists::Must,
            possible_types: Some(shape.type_),
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum BuildError {
    #[error("failed to parse json schema: {0}")]
    InvalidSchema(#[from] SchemaBuildError),
    #[error("cannot process json schema: {0}")]
    SchemaIndex(#[from] SchemaIndexError),
}

/// Resolves a map of possible column names to associated type information. This uses both the
/// `projections` and the `schema` from the `config`. Runs an inferrence on the JSON schema to
/// extract type information about known locations. The returned map will contain several possible
/// aliases for each inferred location, but projections from the config will always take
/// precedence.
///
/// Parsers can use the returned map to lookup type information based on parsed column names.
#[tracing::instrument(skip(config))]
pub fn build_projections(config: &ParseConfig) -> Result<BTreeMap<String, TypeInfo>, BuildError> {
    let schema_uri = url::Url::parse("whatever://placeholder").unwrap();
    let schema_json = if config.schema.is_null() {
        &Value::Bool(true)
    } else {
        &config.schema
    };
    let schema: Schema = json::schema::build::build_schema(schema_uri.clone(), &schema_json)?;
    let mut index = SchemaIndex::new();
    index.add(&schema)?;
    let shape = Shape::infer(&schema, &index);

    let mut results = BTreeMap::new();

    for (pointer, shape, exists) in shape.locations() {
        let target_location = Pointer::from_str(pointer.as_str());
        for resolved_field in derive_field_names(pointer.as_str()) {
            let projection = TypeInfo::from_shape(target_location.clone(), shape, exists);
            results.insert(resolved_field, projection);
        }
    }

    // projections from the configuration always take precedence over those we infer from the
    // schema.
    for (field, pointer) in config.projections.iter() {
        let target_location = Pointer::from_str(pointer.as_ref());
        let projection = if let Some((shape, exists)) = shape.locate(&target_location) {
            TypeInfo::from_shape(target_location, shape, exists)
        } else {
            // This isn't a hard error because there may be files that we can still parse correctly
            // even without knowing the types from the schema. For example, this could be a
            // location that simply allows any valid JSON, and the projection may be only for the
            // sake of putting things into the right shape.
            tracing::warn!(
                field = field.as_str(),
                pointer = pointer.as_ref(),
                "could not locate projection within schema"
            );
            TypeInfo {
                target_location,
                must_exist: false,
                possible_types: None,
            }
        };
        results.insert(field.clone(), projection);
    }

    Ok(results)
}

/// Returns a possibly 0-length collection of field names derived from the given JSON pointer.
/// The field names will represent a variety of possible mappings from fields to the location,
/// which will be used to make a best-effort lookup of columns from a tabular data file.
fn derive_field_names(pointer: &str) -> Vec<String> {
    use doc::ptr::Token;
    if !pointer.is_empty() {
        let split = Pointer::from(pointer)
            .iter()
            .map(|t| match t {
                Token::Index(i) => i.to_string(),
                Token::Property(p) => p.to_string(),
                Token::NextIndex => "-".to_string(),
            })
            .collect::<Vec<_>>();

        let with_underscores = split.iter().fold(String::new(), |mut acc, r| {
            if !acc.is_empty() {
                acc.push('_');
            }
            acc.push_str(r.as_str());
            acc
        });
        let camel_case = split.iter().fold(String::new(), |mut acc, r| {
            if acc.is_empty() {
                acc.push_str(r.as_str());
            } else {
                if let Some(c) = r.chars().next() {
                    acc.push(c.to_ascii_uppercase());
                    acc.extend(r.chars().skip(1));
                }
            }
            acc
        });

        let mut title_case = camel_case
            .chars()
            .take(1)
            .map(|c| c.to_ascii_uppercase())
            .collect::<String>();
        title_case.extend(camel_case.chars().skip(1));

        vec![
            with_underscores,
            camel_case,
            title_case,
            pointer.to_string(),
            pointer[1..].to_string(),
        ]
    } else {
        Vec::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde_json::json;

    macro_rules! map_of{
        ($($key:expr => $val:expr),*) => {{
            let mut m = BTreeMap::new();
            $(
                m.insert($key.into(), $val.into());
             )*
            m
        }}
    }

    #[test]
    fn projections_are_built() {
        let config = ParseConfig {
            projections: map_of!(
                "fieldA" => "/locationa",
                "fieldB" => "/b/loc",
                // Ensure that this projection takes precedence over the generated on
                "BeeLoc" => "/locationa"
            ),
            schema: json!({
                "type": "object",
                "properties": {
                    "locationa": {
                        "type": "integer"
                    },
                    "bee": {
                        "type": "object",
                        "properties": {
                            "loc": { "type": "string" },
                            "rock": {
                                "type": "object",
                                "properties": {
                                    "flower": { "type": "boolean" }
                                }
                            }
                        }
                    }
                },
                "required": ["locationa", "bee"]
            }),
            ..Default::default()
        };
        let result = build_projections(&config).expect("failed to build projectsions");
        insta::assert_debug_snapshot!(result);
    }
}
