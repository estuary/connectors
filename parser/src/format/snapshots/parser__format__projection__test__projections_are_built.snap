---
source: src/format/projection.rs
expression: result

---
{
    "/bee": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: true,
        target_location: [
            Property(
                "bee",
            ),
        ],
    },
    "/bee/loc": TypeInfo {
        possible_types: Some(
            "string",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "loc",
            ),
        ],
    },
    "/bee/rock": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
        ],
    },
    "/bee/rock/flower": TypeInfo {
        possible_types: Some(
            "boolean",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
            Property(
                "flower",
            ),
        ],
    },
    "/locationa": TypeInfo {
        possible_types: Some(
            "integer",
        ),
        must_exist: true,
        target_location: [
            Property(
                "locationa",
            ),
        ],
    },
    "Bee": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: true,
        target_location: [
            Property(
                "bee",
            ),
        ],
    },
    "BeeLoc": TypeInfo {
        possible_types: Some(
            "integer",
        ),
        must_exist: true,
        target_location: [
            Property(
                "locationa",
            ),
        ],
    },
    "BeeRock": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
        ],
    },
    "BeeRockFlower": TypeInfo {
        possible_types: Some(
            "boolean",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
            Property(
                "flower",
            ),
        ],
    },
    "Locationa": TypeInfo {
        possible_types: Some(
            "integer",
        ),
        must_exist: true,
        target_location: [
            Property(
                "locationa",
            ),
        ],
    },
    "bee": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: true,
        target_location: [
            Property(
                "bee",
            ),
        ],
    },
    "bee/loc": TypeInfo {
        possible_types: Some(
            "string",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "loc",
            ),
        ],
    },
    "bee/rock": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
        ],
    },
    "bee/rock/flower": TypeInfo {
        possible_types: Some(
            "boolean",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
            Property(
                "flower",
            ),
        ],
    },
    "beeLoc": TypeInfo {
        possible_types: Some(
            "string",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "loc",
            ),
        ],
    },
    "beeRock": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
        ],
    },
    "beeRockFlower": TypeInfo {
        possible_types: Some(
            "boolean",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
            Property(
                "flower",
            ),
        ],
    },
    "bee_loc": TypeInfo {
        possible_types: Some(
            "string",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "loc",
            ),
        ],
    },
    "bee_rock": TypeInfo {
        possible_types: Some(
            "object",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
        ],
    },
    "bee_rock_flower": TypeInfo {
        possible_types: Some(
            "boolean",
        ),
        must_exist: false,
        target_location: [
            Property(
                "bee",
            ),
            Property(
                "rock",
            ),
            Property(
                "flower",
            ),
        ],
    },
    "fieldA": TypeInfo {
        possible_types: Some(
            "integer",
        ),
        must_exist: true,
        target_location: [
            Property(
                "locationa",
            ),
        ],
    },
    "fieldB": TypeInfo {
        possible_types: None,
        must_exist: false,
        target_location: [
            Property(
                "b",
            ),
            Property(
                "loc",
            ),
        ],
    },
    "locationa": TypeInfo {
        possible_types: Some(
            "integer",
        ),
        must_exist: true,
        target_location: [
            Property(
                "locationa",
            ),
        ],
    },
}
